components/validator/challenger.go

```diff
@@ -24,232 +31,515 @@ type ProofFetcher interface {
-	cfg Config
-	ctx context.Context
+	log    log.Logger
+	cfg    Config
+	ctx    context.Context
+	cancel context.CancelFunc
 
 	callOpts *bind.CallOpts
 	txOpts   *bind.TransactOpts
 
+	l1Client *ethclient.Client
+
 	l2ooContract      *bindings.L2OutputOracle
+	l2ooABI           *abi.ABI
 	colosseumContract *bindings.Colosseum
+	colosseumABI      *abi.ABI
 
 	submissionInterval        *big.Int
 	finalizationPeriodSeconds *big.Int
 	l2BlockTime               *big.Int
 	checkpoint                *big.Int
+
+	txCandidatesChan chan<- txmgr.TxCandidate
+
+	wg sync.WaitGroup
 }
 
-func NewChallenger(ctx context.Context, cfg Config, l log.Logger) (*Challenger, error) {
+func NewChallenger(parentCtx context.Context, cfg Config, l log.Logger, txCandidateChan chan<- txmgr.TxCandidate) (*Challenger, error) {
 	colosseumContract, err := bindings.NewColosseum(cfg.ColosseumAddr, cfg.L1Client)
 	if err != nil {
 		return nil, err
 	}
+
+	colosseumABI, err := bindings.ColosseumMetaData.GetAbi()
+	if err != nil {
+		return nil, err
+	}
+
 	l2ooContract, err := bindings.NewL2OutputOracle(cfg.L2OutputOracleAddr, cfg.L1Client)
 	if err != nil {
 		return nil, err
 	}
 
+	l2ooABI, err := bindings.L2OutputOracleMetaData.GetAbi()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithCancel(parentCtx)
 	callOpts := utils.NewSimpleCallOpts(ctx)
 	submissionInterval, err := l2ooContract.SUBMISSIONINTERVAL(callOpts)
 	if err != nil {
+		cancel()
 		return nil, fmt.Errorf("failed to get submission interval: %w", err)
 	}
 	finalizationPeriodSeconds, err := l2ooContract.FINALIZATIONPERIODSECONDS(callOpts)
 	if err != nil {
+		cancel()
 		return nil, fmt.Errorf("failed to get finalization period seconds: %w", err)
 	}
 	l2BlockTime, err := l2ooContract.L2BLOCKTIME(callOpts)
 	if err != nil {
+		cancel()
 		return nil, fmt.Errorf("failed to get l2 block time: %w", err)
 	}
 
 	return &Challenger{
-		log: l,
-		cfg: cfg,
-		ctx: ctx,
+		log:    l,
+		cfg:    cfg,
+		ctx:    ctx,
+		cancel: cancel,
 
 		callOpts: utils.NewCallOptsWithSender(ctx, cfg.TxManager.From()),
 		txOpts:   utils.NewSimpleTxOpts(ctx, cfg.TxManager.From(), cfg.TxManager.Signer),
 
+		l1Client: cfg.L1Client,
+
 		l2ooContract:      l2ooContract,
+		l2ooABI:           l2ooABI,
 		colosseumContract: colosseumContract,
+		colosseumABI:      colosseumABI,
 
 		submissionInterval:        submissionInterval,
 		finalizationPeriodSeconds: finalizationPeriodSeconds,
 		l2BlockTime:               l2BlockTime,
+
+		txCandidatesChan: txCandidateChan,
 	}, nil
 }
 
-func (c *Challenger) IsChallengeInProgress() (bool, error) {
-	return c.colosseumContract.IsInProgress(c.callOpts)
-}
+func (c *Challenger) Start() error {
+	c.log.Info("start challenger")
+
+	// if checkpoint is behind the latest output index, scan the previous outputs from the checkpoint
+	latestOutputIndex, err := c.l2ooContract.LatestOutputIndex(c.callOpts)
+	if err.Error() == "execution reverted" { // no output submitted
```

Comment:
The handling based on error message wording seems ambiguous, and it would be nice to see some clarification on the reverted behavior.

Reply:
I agree. As you said, it's quite implicit so I changed to check if the `NextOutputIndex` is equal to zero or not.
(If the `NextOutputIndex` is equal to 0, there is no output submitted)